### PR TITLE
Verify donation transaction

### PR DESCRIPTION
**Pull request by mohammadranjbarz**
related https://github.com/Giveth/impact-graph/issues/168

Needed configs

```
BLOCKSCOUT_API_URL=https://blockscout.com/xdai/mainnet/api
ETHERSCAN_MAINNET_API_URL=https://api.etherscan.io/api
ETHERSCAN_ROPSTEN_API_URL=https://api-ropsten.etherscan.io/api
ETHERSCAN_API_KEY=
XDAI_NODE_HTTP_URL=https://dry-small-sound.xdai.quiknode.pro
INFURA_API_KEY=
NUMBER_OF_VERIFY_DONATION_CONCURRENT_JOB=1
VERIFY_DONATION_CRONJOB_EXPRESSION=0 0 * * * *
```

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -111,6 +111,9 @@
       "pre-push": "lint-staged"
     }
   },
+  "standard": {
+    "env": ["mocha"]
```
Comment (mohammadranjbarz) [inline package.json]: Adding this prevents showing invalid tslint warnings in test files
https://stackoverflow.com/a/38667441/4650625

File: provider.ts [hunk 2]
```diff
@@ -1,34 +1,82 @@
...
+
+export function getProvider(networkId: number) {
+  const network = NETWORK_ID_MAP[networkId]
+  if (network === NETWORK_NAMES.XDAI) {
     return new ethers.providers.JsonRpcProvider(
       config.get('XDAI_NODE_HTTP_URL') as string
     )
   }
   // 'https://bsc-dataseed.binance.org/'
-  if (network === 'bsc') {
```
Comment (mohammadranjbarz) [inline provider.ts:27]: Replace these hardcode things with constant variable

File: services/transactionService.ts [hunk 3]
```diff
@@ -1,70 +1,217 @@
-import Web3 from 'web3'
```
Comment (mohammadranjbarz) [inline services/transactionService.ts]: I changed the order of functions to use https://dzone.com/articles/the-stepdown-rule for increase readability of code

**Review by aminlatifi — COMMENTED** [review]
Still we are checking two times and it's not necessary, also added complexity and reduces code readability
We have txHash and nonce, so find a tx that math to any of these two. At the end we will check other values (to, from, amount)

File: services/transactionService.ts [hunk 4]
```diff
@@ -1,70 +1,217 @@
...
+        page,
+        networkId: input.networkId,
+        contractAddress
+      })
+    : await getListOfNormalTransactionsByAddress({
+        address: input.fromAddress,
+        page,
+        networkId: input.networkId
+      })
+  if (isTransactionListEmpty) {
```
Comment (aminlatifi) [inline services/transactionService.ts:75]: What if the first fetched list are all receive transactions (to == userAddress)

**Review by aminlatifi — CHANGES_REQUESTED** [review]
Still we are checking two times and it's not necessary, also added complexity and reduces code readability
We have txHash and nonce, so find a tx that math to any of these two. At the end we will check other values (to, from, amount)

By the way, the code is still complex and hard to follow

File: services/transactionService.ts (hunk 4 above)
Comment (mohammadranjbarz) [reply services/transactionService.ts:75]: In here https://github.com/Giveth/impact-graph/blob/check-donations-with-network-revision2/services/transactionService.ts#L244 
I check if the transaction list is empty ( Doesn't matter our address is source or destination) then we return it true, otherwise, we will check another batch of transactions

**Comment by mohammadranjbarz** [comment]
> Still we are checking two times and it's not necessary, also added complexity and reduces code readability We have txHash and nonce, so find a tx that math to any of these two. At the end we will check other values (to, from, amount)

I would try to simplify it, but it's good if we have another call to talk about It to how make the logic simple

**Comment by MoeNick** [comment]
Would u pls connect it to required issues as well @mohammadranjbarz ?

**Comment by mohammadranjbarz** [comment]
> Would u pls connect it to required issues as well @mohammadranjbarz ?

Sure,  it's connected to #154 
and that PR is connected to https://github.com/Giveth/GIVeconomy/issues/62, https://github.com/Giveth/GIVeconomy/issues/81

**Comment by mohammadranjbarz** [comment]
@aminlatifi 
Could you please review this PR again 
I think it's more readable now  ( I changed getting transaction from etherscan logic especially when finding transaction by nonce, as we talked)

File: services/transactionService.ts [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+export async function getTransactionInfoFromNetwork(
+  input: TransactionDetailInput,
+): Promise<NetworkTransactionInfo> {
+  const { networkId, nonce } = input;
+
+  const web3 = getNetworkWeb3(networkId);
+  const userTransactionsCount = await web3.eth.getTransactionCount(
+    input.fromAddress,
+  );
+  if (nonce && userTransactionsCount <= nonce) {
```
Comment (aminlatifi) [inline services/transactionService.ts]: `nonce` is number and can have 0 value, then this condition will be always false

File: services/transactionService.ts [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+  );
+
+  if (foundTransaction) {
+    return findTransactionByHash({ ...input, txHash: foundTransaction.hash });
+  }
+
+  // userRecentTransactions just includes the transactions that source is our fromAddress
+  // so if the lowest nonce in this array is smaller than the sent nonce we would know that we should not
+  // check latest transactions
+  const smallestNonce: number = userRecentTransactions[
```
Comment (aminlatifi) [inline services/transactionService.ts]: userRecentTransactions length can be zero

File: services/transactionService.ts [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+      errorMessages.TRANSACTION_TO_ADDRESS_IS_DIFFERENT_FROM_SENT_TO_ADDRESS,
+    );
+  }
+
+  if (transaction.from.toLowerCase() !== input.fromAddress.toLowerCase()) {
+    throw new Error(
+      errorMessages.TRANSACTION_FROM_ADDRESS_IS_DIFFERENT_FROM_SENT_FROM_ADDRESS,
+    );
+  }
+  if (transaction.amount !== input.amount) {
```
Comment (aminlatifi) [inline services/transactionService.ts:239]: It's error-prone but cannot fix it by a simple solution. The `amount` data type is defined as `number` and changing it is laborous.

**Review by aminlatifi — CHANGES_REQUESTED** [review]
Thank you @mohammadranjbarz, a big job
Please fix those few issues and after that, it would be OK.

File: services/transactionService.ts (hunk 5 above)
Comment (mohammadranjbarz) [reply services/transactionService.ts]: It has been fixed in #185

File: services/transactionService.ts (hunk 6 above)
Comment (mohammadranjbarz) [reply services/transactionService.ts]: it has been fixed in #185
